commons-core/src/main/scala/com/avsystem/commons/serialization/macroCodecs.scala

```diff
@@ -143,38 +143,43 @@ abstract class FlatSealedHierarchyCodec[T](
   final def readObject(input: ObjectInput): T = {
     val oooFields = new FieldValues(oooFieldNames, oooDeps, typeRepr)
 
-    def read(): T =
-      if (input.hasNext) {
-        val fi = input.nextField()
-        if (fi.fieldName == caseFieldName) {
-          val caseName = readCaseName(fi)
-          caseIndexByName(caseName) match {
-            case -1 => unknownCase(caseName)
-            case idx => readFlatCase(caseName, oooFields, input, caseDeps(idx))
-          }
-        } else if (!oooFields.tryReadField(fi)) {
-          if (caseDependentFieldNames.contains(fi.fieldName)) {
-            if (defaultCaseIdx != -1) {
-              val defaultCaseName = caseNames(defaultCaseIdx)
-              val wrappedInput = new DefaultCaseObjectInput(fi, input, defaultCaseName)
-              readFlatCase(defaultCaseName, oooFields, wrappedInput, caseDeps(defaultCaseIdx))
-            } else {
-              missingCase(fi.fieldName)
-            }
+    def readCase(caseNameField: FieldInput): T = {
+      val caseName = readCaseName(caseNameField)
+      caseIndexByName(caseName) match {
```

Comment:
2 unrelated questions about `caseIndexByName`:
1. Why can't we just use `.indexOf(element, index)`?
2. Does it make sense to use a more lookup-friendly data structure?

Reply:
1. `.indexOf(element, index)` would put the underlying `Array` through implicit conversion to `WrappedArray`. I wanted to be as raw as possible.

2. I'm deliberately using `Array` and linear search here (and also in `FieldValues`) because case classes tend to have small number of fields and sealed hierarchies tend to have small number of case classes. Using a map would probably introduce more overhead.